### PR TITLE
Feat: Show hide series from the legend

### PR DESCRIPTION
**Pull request by wadjih-bencheikh18**
closes : https://github.com/zakodium/react-plot/issues/281

### CODE REVIEW

**Comment by wadjih-bencheikh18** [comment]
@lpatiny about the "eye" do you want me to use a local image?

**Comment by wadjih-bencheikh18** [comment]
@lpatiny I don't know why it's working in local storybook but it' doesn't in deploy preview

File: src/contexts/legendContext.tsx [hunk 1]
```diff
@@ -13,7 +13,7 @@ import type { ActionType, Shape } from '../types';
 interface LegendLabelState {
   id: string;
   label?: string;
-
+  visibility?: boolean;
```
Comment (targos) [inline src/contexts/legendContext.tsx:16]: Please rename to `isVisible` (everywhere). The value should not be optional.

File: src/components/Legend.tsx [hunk 2]
```diff
@@ -160,8 +162,21 @@ export function Legend({
...
             <g
+              onClick={(event) => {
+                onClick?.({ event, id: value.id });
+                if (showHide) {
+                  legendDispatch({
+                    type: 'ADD_LEGEND_LABEL',
+                    payload: {
+                      ...value,
+                      visibility: !value.visibility,
+                    },
```
Comment (targos) [inline src/components/Legend.tsx:173]: Please add a new action type called `TOGGLE_VISIBILITY`, which only takes the series id and the reducer can just update the boolean.
This part would be changed to : `legendDispatch({ type: 'TOGGLE_VISIBILITY', payload: { seriesId: value.id } })`

File: src/components/Legend.tsx [hunk 3]
```diff
@@ -187,9 +202,19 @@ export function Legend({
...
             onClick={(event) => {
               onClick?.({ event, id: value.id });
+              if (showHide) {
+                legendDispatch({
+                  type: 'ADD_LEGEND_LABEL',
+                  payload: {
+                    ...value,
+                    visibility: !value.visibility,
+                  },
+                });
```
Comment (targos) [inline src/components/Legend.tsx:212]: ditto

File: src/components/BarSeries.tsx [hunk 4]
```diff
@@ -37,22 +37,26 @@ export function BarSeries(props: BarSeriesProps) {
       hidden: !displayMarker,
     };
   }, [color, displayMarker, figure]);
-
+  const visibility = useMemo(() => {
+    const value = legendState.labels.find((label) => label.id === id);
+    return value ? value.visibility : true;
+  }, [id, legendState.labels]);
```
Comment (targos) [inline src/components/BarSeries.tsx:43]: This same code is used in all series components. Please refactor to a custom hook. `const isVisible = useIsSeriesVisible(id)`

File: src/components/BarSeries.tsx [hunk 5]
```diff
@@ -37,22 +37,26 @@ export function BarSeries(props: BarSeriesProps) {
...
   useEffect(() => {
     legendDispatch({
       type: 'ADD_LEGEND_LABEL',
       payload: {
         id,
+        visibility,
```
Comment (targos) [inline src/components/BarSeries.tsx:49]: We do not need to pass `isVisible` here. The default value should be `true` when the label is added.

**Review by targos — DISMISSED** [review]

File: src/components/BarSeries.tsx (hunk 5 above)
Comment (wadjih-bencheikh18) [reply src/components/BarSeries.tsx:49]: this dispatch is used to update every moment so if i set it to true the TOGGLE_VISIBILITY won't work correctly

File: src/contexts/legendContext.tsx (hunk 1 above)
Comment (wadjih-bencheikh18) [reply src/contexts/legendContext.tsx]: fixed

File: src/components/BarSeries.tsx (hunk 5 above)
Comment (targos) [reply src/components/BarSeries.tsx]: Only TOGGLE_VISIBILITY should modify `isVisible` you can keep the current state value when `ADD_LEGEND_LABEL` is handled.

Comment (wadjih-bencheikh18) [reply src/components/BarSeries.tsx]: same problem

File: src/components/LineSeries.tsx [hunk 6]
```diff
@@ -29,16 +30,16 @@ export function LineSeries(props: LineSeriesProps) {
     ...otherProps
   } = props;
   const lineStyle = functionalStyle({}, OldLineStyle, { id });
+  const isVisible = useIsSeriesVisible(id);
   useEffect(() => {
```
Comment (wadjih-bencheikh18) [inline src/components/LineSeries.tsx:34]: @targos you can see here
every moment we're adding and removing legend label
(it's not working now)

File: src/contexts/legendContext.tsx [hunk 7]
```diff
@@ -33,7 +33,8 @@ interface LegendState {
 
 type LegendActions =
   | ActionType<'ADD_LEGEND_LABEL', LegendLabelState>
```
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/contexts/legendContext.tsx]: added , but still don't work

File: src/components/LineSeries.tsx (hunk 6 above)
Comment (wadjih-bencheikh18) [reply src/components/LineSeries.tsx:34]: we're changing LegendLabels state

File: src/contexts/legendContext.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (wadjih-bencheikh18) [reply src/contexts/legendContext.tsx]: fixed

**Comment by stropitek** [comment]
@wadjih-bencheikh18 there are conflicts again 🥲

**Comment by wadjih-bencheikh18** [comment]
> In the control and the showHide stories, the showHide prop control does not seem to have any effect

as i mentioned 
i don't know why it's working in local perfectly but here no

**Comment by stropitek** [comment]
I couldn't understand why the netlify build is different from a local build. I implemented a workaround, it seems like the problem is gone if you don't extend the legend's props with other props.

**Pull request merged**